src/com/amazon/ionelement/api/IonElement.kt

```diff
@@ -94,19 +94,57 @@ interface IonElement {
 
     /** Converts the current element to Ion text. */
     override fun toString(): String
-}
 
+    /** Returns a shallow copy of the current node with the specified additional annotations. */
+    fun withAnnotations(vararg additionalAnnotations: String): IonElement
+
+    /** Returns a shallow copy of the current node with the specified additional annotations. */
+    fun withAnnotations(additionalAnnotations: Iterable<String>): IonElement
+
+    /** Returns a shallow copy of the current node with all annotations removed. */
+    fun withoutAnnotations(): IonElement
+
+    /**
+     * Returns a shallow copy of the current node with the specified additional metadata, overwriting any metas
+     * that already exist with the same keys.
+     */
+    fun withMetas(additionalMetas: MetaContainer): IonElement
+
+    /**
+     * Returns a shallow copy of the current node with the specified additional meta, overwriting any meta
+     * that previously existed with the same key.
+     *
+     * When adding multiple metas, consider [withMetas] instead.
+     */
+    fun withMeta(key: String, value: Any): IonElement
+
+    /** Returns a shallow copy of the current node without any metadata. */
+    fun withoutMetas(): IonElement
+}
 
 /** Represents a Ion bool. */
 interface BoolElement : IonElement {
     val booleanValue: Boolean
     override fun copy(annotations: List<String>, metas: MetaContainer): BoolElement
+    override fun withAnnotations(vararg additionalAnnotations: String): BoolElement
+    override fun withAnnotations(additionalAnnotations: Iterable<String>): BoolElement
+    override fun withoutAnnotations(): BoolElement
+    override fun withMetas(additionalMetas: MetaContainer): BoolElement
+    override fun withMeta(key: String, value: Any): BoolElement
+    override fun withoutMetas(): BoolElement
 }
```

Comment:
Is there any place in e.g. IonElement for a comment detailing why you've chosen repetition over generics? We had an offline conversation and you had a good reason charting the best path between several less desirable possibilities so it would be nice to capture some of that reasoning for posterity.

Reply:
I've added a comment in the `IonElement` interface, and I'll add some info here.

#### Generics
I did test out an approach using generics (`IonElement<T: IonElement<T>>`) that would allow child interfaces and subclasses to automatically get appropriately typed functions to implement without all of the repetition and boiler plate, but that approach requires the type parameter in every sub-interface (annoying perhaps, but not a blocker), and we still run into a conflict for the classes which extend `AnyElementBase` and implement their respective element-type interfaces. 

#### De-Reification
I also tested out an approach where, I modified the existing `with*` functions so that the type parameter is not inline/reified, like this:
```kotlin
fun <T: IonElement> T._withAnnotations(vararg additionalAnnotations: String): T =
    when {
        additionalAnnotations.isEmpty() -> this
        else -> copy(annotations = this.annotations + additionalAnnotations) as T
    }
```
This approach is less desirable because the compiler can no longer infer that this is a safe cast (not necessarily a blocker), and because java callers would have to call it as a static utility method because of how extension functions are exposed to Java. Eg.:
```java
import static com.amazon.ionelement.api.IonElementExtensionsKt.withAnnotations;
import static com.amazon.ionelement.api.IonElementExtensionsKt.withMeta;

IonElement b = withMeta(withAnnotations(a, "foo"), "bar", 42);
```

#### Explicit covariant returns in every sub-interface
The approach I used in this PR has a lot of repetition/boilerplate, but it allows Java code to call these functions ergonomically.
```java
IonElement b = a.withAnnotations("foo").withMeta("bar", 42);
```